### ISSUE TITLE
[Bug]: log scale on unitized data

### ISSUE THREAD

**Issue by story645** [matplotlib/matplotlib#20979]
### Bug summary

Not totally sure if this is a bug, or a feature request, but for now should we maybe add a warning on trying to do log scale w/ dates & categories - i.e. non-float datatypes that the units system converts to floats but the resulting figure strips out all the unit information because it does the math on the underlying floats and doesn't convert back to the units. 

### Code for reproduction

```python
fig, ax = plt.subplots()
ax.plot(['a','b','c'], 
        [datetime(1999,12,1), datetime(2000,1,1), 
        datetime(2020, 10,1)])
ax.set_xscale('log')
ax.set_yscale('log')
```

### Actual outcome

![unit](https://user-images.githubusercontent.com/1300499/131887631-6a4409cf-c72c-43da-b742-aa72e5db43c8.png)

### Expected outcome

probably for the y-axis to still be labled in dates, and the x axis to be labeled in categories (which admittedly is impossible to do in this case)

**Comment by anntzer** [comment]
Interaction between scales and units is a bit underspecified, see e.g. https://github.com/matplotlib/matplotlib/pull/20340 and linked issues.

**Comment by story645** [comment]
Yeah this should at the least throw a warning, possibly error out,

**Comment by sanrishi** [comment]
Hi @story645 and @anntzer,

I've been looking into this and would love to open a PR to implement a guard. I traced the code paths to see exactly how units and scales interact here, and I found that this behavior actually manifests in two directions (forward and reverse).

Here is the proposed approach to catch both cases:

### 1. The Forward Path (Plot Units ➔ Set Scale)
When ax.set_xscale('log') is called after plotting, it hits Axis._set_axes_scale (lib/matplotlib/axis.py:790). This unconditionally applies the scale's default locators and formatters (e.g., LogLocator and LogFormatter), completely overwriting the unit converter's AxisInfo.

Proposed Guard: Inject a check at the top of Axis._set_axes_scale to see if self.have_units() is True and the incoming scale is non-linear ("log", "symlog", "asinh", "logit").

### 2. The Reverse Path (Set Scale ➔ Plot Units)
If a user sets the scale first and then plots dates or categories, Axis.update_units (lib/matplotlib/axis.py:1754) is triggered. This sets the converter, but the underlying scale transform remains non-linear, which conceptually breaks down (especially for categories mapping to 0, which log clips).

Injecting a symmetric check inside Axis.update_units. If a converter is successfully acquired for the new data, and self.get_scale() is already one of the non-linear scales, trigger the same guard.

(To avoid duplication, both of these checks could delegate to a single shared private helper method).

Would you prefer this guard to emit a UserWarning (e.g., via _api.warn_external), or should it be a hard failure (ValueError / TypeError) when this incompatible intersection is detected?

Let me know what you think, and I'll get a PR drafted!

**Comment by sanrishi** [comment]
Hey @story645, fair catch! 😅 I won't lie, I used a local CLI(as i mention on my PRs) to help me grep through the axis.py and scale.py for finding best approach because the Matplotlib codebase is massive, and I used an LLM to help format my notes so I didn't post a giant wall of text.

But I am actually working through the implementation locally, and the logic for the forward/reverse paths is what I've been testing. To prove I'm not just copy-pasting, here is the actual approach I'm taking for the PR.

1. To avoid duplicating the check in both _set_axes_scale and update_units, I'm writing a private helper method in Axis that looks something like this:

```Python
def _check_scale_units_collision(self, scale_name):
    # Check if we have units AND a non-linear scale that strips them
    incompatible_scales = ['log', 'symlog', 'logit', 'asinh']
    if self.have_units() and scale_name in incompatible_scales:
        _api.warn_external(
            f"Applying a '{scale_name}' scale to an axis with units "
            "will override the unit formatting. Consider formatting ticks manually."
        )
```
Then, I'm just injecting that into the two paths we discussed:

2. At the top of _set_axes_scale: self._check_scale_units_collision(value)

Inside update_units (after the converter is set): 
```python
self._check_scale_units_collision(self.get_scale())
```
I currently have this set up locally with a UserWarning. Is this approach looking right path forward to you?

**Comment by story645** [comment]
> Hey @story645, fair catch! 😅 I won't lie

Your solution to being told to not use AI is to use more AI? Given that I just chatted with you for an hour or so, I know this is not how you naturally write.

**Comment by sanrishi** [comment]
Not this time @story645 i am just nervous in meeting but most of the time i chats unprofessionally sorry  😅

**Comment by sanrishi** [comment]
How's my approach looking is there anything which you worry of? 
let me know!

**Comment by story645** [comment]
>  goes little unproffesinal this time

The red flag is how you're using an idiomatic colloquialism. And that your proposal is roughly the one we discussed on the call as not being a good approach because it doesn't account for third party units that would allow for log transforms.

**Comment by sanrishi** [comment]
>Hey @story645, fair catch! 

Sorry @story645 i goes outside and by taking some fresh air now i realizing how idiot i am and does not have faith on my own😭
but i swear from now i not use ai to refine my reply anymore

**Comment by sanrishi** [comment]
>And that your proposal is roughly the one we discussed on the call as not being a good approach because it doesn't account for third party units that would allow for log transforms.

Thanks for clarification!
I now understand my if self.have_units() and scale_name in incompatible_scales is too broad and blocking ALL unit types from using log scale

i will in touch with you after taking some look on codebase!

**Comment by anntzer** [comment]
[Given what I read here I'll just ignore the ping above (though @story645 should feel free to ping me again if needed).]

**Comment by sanrishi** [comment]
You were totally right. Because ConversionInterface doesn't have a built-in way for a converter to declare

So now we have two options to tackle the issue without breaking third-party libraries,
We can go for short term fix by checking if the active converter is one of the known "semantic" ones (like StrCategoryConverter or _SwitchableDateConverter). This safely ignores numeric wrappers like Pint, but feels a bit hardcoded.

Or add a  new optional method/property to the base ConversionInterface (something like supports_scale(scale_name)) so any third-party library can declare if it works with non-linear scales.

Personally, I lean towards Option 2. It seems much better for Matplotlib's long-term stability since it avoids hardcoding specific converter classes directly into the core axis.py logic.
Which one do you prefer? @story645 @anntzer

**Comment by anntzer** [comment]
Please stop pinging me here.

**Comment by story645** [comment]
> You were totally right. Because ConversionInterface doesn't have a built-in way for a converter to declare

Right about what? This is an incomplete sentence.

> This safely ignores numeric wrappers like Pint, but feels a bit hardcoded.

Both solutions are hard coded, they're just trading off where the declaration of support happens/is checked. I told you to look at the paleo climate library to see how they handle dates on a log scale b/c 

Also again, I suspect that these proposals are AI slop b/c frankly the writing style in 'em looks more like mine then yours.

Please understand that your use of AI is counter-productive to your goal of becoming a member of our developer community - while it may not be your intent, to me it reads like you are feeding easy looking problems into a chat bot and throwing whatever solutions it spits back at us to see if anything sticks. 

I am sympathetic - I have been desperate and in over my head many times - which is also why I keep lecturing you on this. Take a few steps back and try to see through #31181 with your own skills at your own pace. Afterwards, we can revisit this topic.

I'm also going to warn you again that you will  get banned from posting in the repo if you keep using AI as you have been.

**Comment by sanrishi** [comment]
Due to tackling so much issues at a time i litterely forget that you referenced pyleoclim(deep-time dates on a log scale) I blindly followed my own analysis sorry for that!

I am reviewing #31181 PR without rushing on it
Thanks for guidance!
Ques: are my original english bad ? i want to clear this up so next time i will not use ai!

**Comment by story645** [comment]
> I blindly followed my own analysis sorry for that!

No, you used AI. At least for writing up the proposal, but I have no confidence that you understand the proposal you posted b/c of your heavy use of AI.
 

> i want to clear this up so next time i will not use ai!

Do not use AI b/c your use of AI violates our policy of [genuine interaction](https://matplotlib.org/devdocs/devel/contribute.html#restrictions-on-generative-ai-usage).

**Comment by sanrishi** [comment]
> but I have no confidence that you understand the proposal you posted b/c of your heavy use of AI.

@story645 That's not the case I researched on the issue, that first proposal to you is not random texts from ai 

I was thinking that you angry to me due to the use of ai on refining my proposal but I does not have any idea that you considering me this dumb, but I want to clarify you that 

I am game developer and was worked for a unreal engine project so in that transforms and annotations are compulsory to be known there!

Because my English is weak thats why I asked you "are my English bad(is bad english makes bad impression in your community)?

I just little confused previous days, Never mind if anything looking arguable to you

**Comment by story645** [comment]
> is bad english makes bad impression in your community?

No, it's how you've been using AI that's been  making a bad impression.

**Comment by story645** [comment]
Also @sanrishi, I'm not angry at you, I just want you to be more careful in your use of AI.

I have been having this conversation in public like this b/c your behavior is common around GSOC time (so again, I get it - I was a gsoc mentee twice) and I want to communicate to you and other applicants that the most effective approach to get positive attention from the maintainers is to sincerely focus on one thing, clearly show that you're making a good faith effort to understand that thing (AI use undermines this), and see it through.

**Comment by sanrishi** [comment]
I went through pyleoclim as you suggested and seing how they bypassing matplotlib units entirely to tackle log-scaling unitized issue

As you asked my have_units() approach will annoy third party libraries so i decided if we specifically target DateConverter and StrCategoryConverter to handle cases where the scale is set before the data is plotted?

This way, we catch the collision regardless of the user's code order, without spamming the warning during every draw cycle!

**Comment by story645** [comment]
> so i decided if we specifically target DateConverter and StrCategoryConverter to handle cases where the scale is set before the data is plotted?

Scale is set by any convertor picked up by the registry. Date and Str function the same as a third party, they're just internal.

**Comment by sanrishi** [comment]
Thanks for the clarification! @story645 as hardcoding isinstance checks completely breaks the abstraction I was just looking at ConversionInterface.axisinfo(unit, axis)

I noticed that when the axis object is passed in, both DateConverter and StrCategoryConverter that ignores it and blindly return their default formatters regardless of axis.get_scale(). This is looking like root problem 

But if we can't isinstance-check, how do we distinguish a converter that actually handles log scales from one that doesn't? Is the right move to add a property on ConversionInterface (like supports_log_scale = True by default, which Date/Str converters can override to False), or does the warning logic live somewhere else entirely?

**Comment by story645** [comment]
> I noticed that when the axis object is passed in, both DateConverter and StrCategoryConverter that ignores it and blindly return their default formatters regardless of axis.get_scale(). This is looking like root problem

Which means this is something that you should have noticed from the first time you proposed this, so this makes me think this post doesn't have the previous context cause it's AI generated. And you were told to focus on your already open PRs.

Because you have failed to follow instructions/heed warnings numerous times, you're being banned from the repo for a week. Given maintainer support for this ban, it also means we are unlikely to select you for GSOC.

If you choose to come back after the ban, please be more careful about following instructions. You are welcome to reach out on incubator if you have questions.
